compiler/src/diagnostics/comments.re

```diff
@@ -0,0 +1,206 @@
+open Grain_parsing;
+open Grain_typed;
+
+// Attributes in a comment are prefixed with `@` symbol, such as `@param`
+module Attribute = {
+  type attr_name = string;
+  type attr_desc = string;
+  // The `attr_type` always starts as `None` and is applied later by something like Graindoc
+  type attr_type = option(string);
+
+  type t =
+    | Param({
+        attr_name,
+        attr_type,
+        attr_desc,
+      })
+    | Returns({
+        attr_desc,
+        attr_type,
+      })
+    | Module({
+        attr_name,
+        attr_desc,
+      })
+    // Currently only accepts single-line examples
+    | Example({attr_desc})
+    | Section({
+        attr_name,
+        attr_desc,
+      });
+
+  let extract = comment => {
+    let attrs = ref([]);
+    // https://regexper.com/#%5E%40%28param%7Creturns%29%28%5B%20%5Ct%5D%28%5B%5E%3A%5D%2B%29%3A%29%3F%5B%20%5Ct%5D%28.*%29%24
+    let re =
+      Str.regexp(
+        "^@\\(param\\|returns\\|module\\|example\\|section\\)\\([ \t]\\([^:]+\\):\\)?[ \t]\\(.*\\)$",
```

Comment:
I'd make a tiny tweak here to consume more whitespace:
```suggestion
        "^@\\(param\\|returns\\|module\\|example\\|section\\)\\([ \t]+\\([^:]+\\):\\)?[ \t]+\\(.*\\)$",
```

Would we want to match word characters instead of the explicit param/returns/module/example/section? That'd have the benefit of being able to error/warn on unknown attributes, which would be handy and would also prevent an unknown one from showing up in a description.

Reply:
Yeah, I like that idea! I'll make that change in this PR

Reply:
Updated. Also added better error messages!